File: packages/python/elizaos/features/basic_capabilities/providers/attachments.py
from __future__ import annotations

from typing import TYPE_CHECKING

from elizaos.generated.spec_helpers import require_provider_spec
from elizaos.types import Provider, ProviderResult

if TYPE_CHECKING:
    from elizaos.types import IAgentRuntime, Memory, State

# Get text content from centralized specs
_spec = require_provider_spec("ATTACHMENTS")


def format_attachment(attachment: dict[str, str]) -> str:
    att_type = attachment.get("type", "unknown")
    url = attachment.get("url", "")
    title = attachment.get("title", "")
    description = attachment.get("description", "")

    parts = [f"- Type: {att_type}"]
    if title:
        parts.append(f"  Title: {title}")
    if description:
        parts.append(f"  Description: {description}")
    if url:
        parts.append(f"  URL: {url}")

    return "\n".join(parts)


async def get_attachments(
    runtime: IAgentRuntime,
    message: Memory,
    state: State | None = None,
) -> ProviderResult:
    attachments: list[dict[str, str]] = []

    if message.content and hasattr(message.content, "attachments"):
        raw_attachments = message.content.attachments or []
        for att in raw_attachments:
            if isinstance(att, dict):
                attachments.append(att)
            elif hasattr(att, "__dict__"):
                attachments.append(att.__dict__)

    if not attachments:
        return ProviderResult(
            text="",
            values={"hasAttachments": False, "attachmentCount": 0},
            data={"attachments": []},
        )

    formatted_attachments = "\n".join(format_attachment(att) for att in attachments)

    text = f"# Attachments ({len(attachments)})\n{formatted_attachments}"

    return ProviderResult(
        text=text,
        values={
            "hasAttachments": True,
            "attachmentCount": len(attachments),
            "attachmentTypes": list({att.get("type", "unknown") for att in attachments}),
        },
        data={
            "attachments": attachments,
        },
    )


attachments_provider = Provider(
    name=_spec["name"],
    description=_spec["description"],
    get=get_attachments,
    dynamic=_spec.get("dynamic", True),
)

File: packages/typescript/src/features/basic-capabilities/providers/attachments.ts
import { requireProviderSpec } from "../../../generated/spec-helpers.ts";
import type {
	IAgentRuntime,
	Media,
	Memory,
	Provider,
} from "../../../types/index.ts";
import { addHeader } from "../../../utils.ts";

// Get text content from centralized specs
const spec = requireProviderSpec("ATTACHMENTS");
const MAX_VISIBLE_ATTACHMENTS = 3;

type AttachmentWithCreatedAt = Media & {
	_createdAt?: number;
};

function mergeConversationAttachments(
	message: Memory,
	recentMessages: Memory[] | null | undefined,
): AttachmentWithCreatedAt[] {
	const attachmentsById = new Map<string, AttachmentWithCreatedAt>();

	const rememberAttachment = (attachment: Media, createdAt: number): void => {
		const existing = attachmentsById.get(attachment.id);
		if (existing && (existing._createdAt ?? 0) >= createdAt) {
			return;
		}
		attachmentsById.set(attachment.id, {
			...attachment,
			_createdAt: createdAt,
		});
	};

	for (const attachment of message.content.attachments ?? []) {
		rememberAttachment(attachment, message.createdAt ?? Date.now());
	}

	for (const recentMessage of recentMessages ?? []) {
		for (const attachment of recentMessage.content.attachments ?? []) {
			rememberAttachment(attachment, recentMessage.createdAt ?? Date.now());
		}
	}

	return Array.from(attachmentsById.values()).sort(
		(left, right) => (right._createdAt ?? 0) - (left._createdAt ?? 0),
	);
}

/**
 * Provides a list of attachments in the current conversation.
 * @param {IAgentRuntime} runtime - The agent runtime object.
 * @param {Memory} message - The message memory object.
 * @returns {Object} The attachments values, data, and text.
 */
/**
 * Provides a list of attachments sent during the current conversation, including names, descriptions, and summaries.
 * @type {Provider}
 * @property {string} name - The name of the provider (ATTACHMENTS).
 * @property {string} description - Description of the provider.
 * @property {boolean} dynamic - Indicates if the provider is dynamic.
 * @property {function} get - Asynchronous function that retrieves attachments based on the runtime and message provided.
 * @param {IAgentRuntime} runtime - The runtime environment for the agent.
 * @param {Memory} message - The message object containing content and attachments.
 * @returns {Object} An object containing values, data, and text about the attachments retrieved.
 */
export const attachmentsProvider: Provider = {
	name: spec.name,
	description: spec.description,
	dynamic: spec.dynamic ?? true,
	get: async (runtime: IAgentRuntime, message: Memory) => {
		const { roomId } = message;
		const conversationLength = runtime.getConversationLength();

		const recentMessagesData = await runtime.getMemories({
			roomId,
			limit: conversationLength,
			unique: false,
			tableName: "messages",
		});

		const allAttachments = mergeConversationAttachments(
			message,
			Array.isArray(recentMessagesData) ? recentMessagesData : [],
		);
		const visibleAttachments = allAttachments.slice(0, MAX_VISIBLE_ATTACHMENTS);
		const omittedCount = Math.max(
			0,
			allAttachments.length - visibleAttachments.length,
		);

		// Format attachments for display
		const formattedAttachments = visibleAttachments
			.map(
				(attachment) =>
					`ID: ${attachment.id}
    Name: ${attachment.title}
    URL: ${attachment.url}
    Type: ${attachment.source}
    Content Type: ${attachment.contentType ?? "unknown"}
    Stored Content: ${
			attachment.text || attachment.description
				? "available via READ_ATTACHMENT"
				: "none"
		}
    `,
			)
			.join("\n");
		const omissionNotice =
			omittedCount > 0
				? `Showing the ${visibleAttachments.length} most recent attachments. ${omittedCount} older attachment${omittedCount === 1 ? "" : "s"} omitted from context; use READ_ATTACHMENT to inspect one.`
				: "";

		// Create formatted text with header
		const text =
			formattedAttachments && formattedAttachments.length > 0
				? addHeader(
						"# Attachments",
						[formattedAttachments, omissionNotice].filter(Boolean).join("\n\n"),
					)
				: "";

		const values = {
			attachments: text,
		};
		const data = {
			attachments: allAttachments,
			visibleAttachments,
			omittedCount,
		};

		return {
			values,
			data,
			text,
		};
	},
};
